Specialize ARRAY_SORT for simple types

kaikalur opened the pull request:
Currently, ARRAY_SORT with default comparator is implemented in a generic fashion making it too expensive. So we will be specializing it for numeric types (int/bigint/double/real) - the most common use cases.

Test Plan: tests exist and also added new tests

```
== NO RELEASE NOTE ==
```

agrawaldevesh commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraySortFunction.java:
```diff
@@ -115,6 +117,56 @@ public int compare(Integer p1, Integer p2)
+        Arrays.sort(values);
+
+        BlockBuilder blockBuilder = BIGINT.createBlockBuilder(null, arrayLength);
+        if (array.mayHaveNull()) {
```
Can you set a flag on L135 instead of the "maybe" part -- I think we want to be more sure on the output. Further, here you should check for NOT having nulls.

kaikalur replied:
This is a standard pattern for handling nulls efficiently. Look at other array operators

agrawaldevesh replied:
Ok, but I am wondering if you should be using the negation of this condition: the loop in 152 assumes no nulls, so maybe !array.maybeHaveNull() ?

agrawaldevesh reviewed (changes requested):
I smell a bug :-)

agrawaldevesh commented:
Would be great to fill out the PR description as well and the wins achieved.

agrawaldevesh commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraySortFunction.java:
```diff
@@ -115,6 +117,56 @@ public int compare(Integer p1, Integer p2)
     }
 
+    @SqlType("array(bigint)")
+    public static Block bigintSort(@SqlType("array(bigint)") Block array)
```
Should we proactively specialize other types as well ?

agrawaldevesh replied:
eg. See the comment TODO on L73 and L92 -- can we do anything about those ?

kaikalur replied:
Yeah - will do also for double (hence draft lol)

kaikalur commented:
> Would be great to fill out the PR description as well and the wins achieved.

this is a draft not ready for review

kaikalur commented:
> I smell a bug :-)

Yeah I switched null and not null. will fix it. I just created the draft pr for you to see the idea.

kaikalur commented:
> > I smell a bug :-)
> 
> Yeah I switched null and not null. will fix it. I just created the draft pr for you to see the idea.

cleaned up completely.

agrawaldevesh commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -765,27 +766,27 @@ public void testArrayAgg()
     @Test
     public void testReduceAgg()
     {
-        assertQuery(
```
I did not follow why this test needs to be changed ?

kaikalur replied:
> I did not follow why this test needs to be changed ?

Oops fixed now

agrawaldevesh commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -7016,6 +7017,42 @@ public void testCaseToMapOptimization()
+        MaterializedResult result = computeActual("select array_sort(array[5,6,4,null,3,7, cast(9223372036854775807 as bigint)])");
+        assertSorted(result.getMaterializedRows().get(0).getFields());
+
+        result = computeActual("select array_sort(array[5.4,6.1,4.9,null,4.5,7.8,5.2])");
```
Can we test that the result is indeed a sorted representation of the original: ie same size and set of values. Also can we test a case of all null values and length less than 2.

kaikalur replied:
we can't do that for non-integer types

kaikalur replied:
there were other tests for this function before - we just specialized it. So no need to add other tests.

pranjalssh commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -7016,6 +7017,42 @@ public void testCaseToMapOptimization()
     public void testLikePrefix()
     {
-        assertQuery("select x like 'abc%' from (values 'abc', 'def', 'bcd') T(x)");
+        assertNotEquals(computeActual("EXPLAIN(TYPE DISTRIBUTED) select x like 'abc%' from (values 'abc', 'def', 'bcd') T(x)").getOnlyValue().toString().toLowerCase(Locale.ROOT).indexOf("substr"), -1);
```
Unrelated test?

kaikalur replied:
Removed

pranjalssh commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraySortFunction.java:
```diff
@@ -115,6 +120,134 @@ public int compare(Integer p1, Integer p2)
+
+        Arrays.sort(values, 0, values.length - nulls);
+
+        BlockBuilder blockBuilder = BIGINT.createBlockBuilder(null, arrayLength);
```
Is it possible to not create extra array values, and reuse the values array in block builder?

kaikalur replied:
> Is it possible to not create extra array values, and reuse the values array in block builder?

Not really. That's internal and can vary so we want to follow the Block API properly. This is just an opportunistic specialization for primitive types.

pranjalssh reviewed (approved):
Looks good
